Energy Consumption and Prediction

Dataset Description:
This dataset contains measurements of electric power consumption in a household, recorded over a period of time. The data includes various attributes that represent different aspects of the household's energy usage. It is suitable for analyzing power consumption trends, identifying usage patterns, and predicting future consumption levels.

Dataset:
Attributes:

Date: The date of the measurement (in DD-MM-YYYY format).
Time: The time of the measurement (in HH:MM:SS format).
Global_active_power (kW): The total active power consumed by the household in kilowatts (kW). This refers to the power used by all electrical devices in the house.
Global_reactive_power (kVAR): The reactive power consumed, measured in kilovolt-amperes reactive (kVAR), which represents the power lost due to inductance and capacitance.
Voltage (V): The household voltage at the time of the measurement.
Global_intensity (A): The total current drawn by the household in amperes (A).
Sub_metering_1 (watt-hour of active energy): Energy consumption from the kitchen, including appliances like the dishwasher, oven, and microwave.
Sub_metering_2 (watt-hour of active energy): Energy consumption from the laundry area, including washing machines, dryers, and refrigerators.
Sub_metering_3 (watt-hour of active energy): Energy consumption from electric water heaters and air conditioners.

Dataset Link: 
https://drive.google.com/file/d/1bvaXJJqNObOCkX-i475BNxpidk024pyx/view?usp=sharing


Importing libraries

In [3]:
import numpy as np
import pandas as pd


Loading the dataset

In [63]:
df=pd.read_csv('household_power_consumption.txt',sep=";")
# Prints the dataset
df

C:\Users\Ayush\AppData\Local\Temp\ipykernel_13360\3750999053.py:1: DtypeWarning: Columns (2,3,4,5,6,7) have mixed types. Specify dtype option on import or set low_memory=False.
  df=pd.read_csv('household_power_consumption.txt',sep=";")


Date      Time Global_active_power Global_reactive_power  \
0        16/12/2006  17:24:00               4.216                 0.418   
1        16/12/2006  17:25:00               5.360                 0.436   
2        16/12/2006  17:26:00               5.374                 0.498   
3        16/12/2006  17:27:00               5.388                 0.502   
4        16/12/2006  17:28:00               3.666                 0.528   
...             ...       ...                 ...                   ...   
2075254  26/11/2010  20:58:00               0.946                   0.0   
2075255  26/11/2010  20:59:00               0.944                   0.0   
2075256  26/11/2010  21:00:00               0.938                   0.0   
2075257  26/11/2010  21:01:00               0.934                   0.0   
2075258  26/11/2010  21:02:00               0.932                   0.0   

         Voltage Global_intensity Sub_metering_1 Sub_metering_2  \
0        234.840           18.400          0.000          1.000   
1        233.630           23.000          0.000          1.000   
2        233.290           23.000          0.000          2.000   
3        233.740           23.000          0.000          1.000   
4        235.680           15.800          0.000          1.000   
...          ...              ...            ...            ...   
2075254   240.43              4.0            0.0            0.0   
2075255    240.0              4.0            0.0            0.0   
2075256   239.82              3.8            0.0            0.0   
2075257    239.7              3.8            0.0            0.0   
2075258   239.55              3.8            0.0            0.0   

         Sub_metering_3  
0                  17.0  
1                  16.0  
2                  17.0  
3                  17.0  
4                  17.0  
...                 ...  
2075254             0.0  
2075255             0.0  
2075256             0.0  
2075257             0.0  
2075258             0.0  

[2075259 rows x 9 columns]

In [14]:
# This shows the top 5 records of the dataset
df.head()

Date      Time Global_active_power Global_reactive_power  Voltage  \
0  16/12/2006  17:24:00               4.216                 0.418  234.840   
1  16/12/2006  17:25:00               5.360                 0.436  233.630   
2  16/12/2006  17:26:00               5.374                 0.498  233.290   
3  16/12/2006  17:27:00               5.388                 0.502  233.740   
4  16/12/2006  17:28:00               3.666                 0.528  235.680   

  Global_intensity Sub_metering_1 Sub_metering_2  Sub_metering_3  
0           18.400          0.000          1.000            17.0  
1           23.000          0.000          1.000            16.0  
2           23.000          0.000          2.000            17.0  
3           23.000          0.000          1.000            17.0  
4           15.800          0.000          1.000            17.0

In [18]:
# This shows the last 5 records of the dataset
df.tail()

Date      Time Global_active_power Global_reactive_power  \
2075254  26/11/2010  20:58:00               0.946                   0.0   
2075255  26/11/2010  20:59:00               0.944                   0.0   
2075256  26/11/2010  21:00:00               0.938                   0.0   
2075257  26/11/2010  21:01:00               0.934                   0.0   
2075258  26/11/2010  21:02:00               0.932                   0.0   

        Voltage Global_intensity Sub_metering_1 Sub_metering_2  Sub_metering_3  
2075254  240.43              4.0            0.0            0.0             0.0  
2075255   240.0              4.0            0.0            0.0             0.0  
2075256  239.82              3.8            0.0            0.0             0.0  
2075257   239.7              3.8            0.0            0.0             0.0  
2075258  239.55              3.8            0.0            0.0             0.0

In [97]:
# summary statistics for numerical columns
# This provides basic statistical measures such as mean, min, max, standard deviation, and percentiles for each numeric column.
df.describe()

Global_active_power  Global_reactive_power       Voltage  \
count         2.075259e+06           2.075259e+06  2.075259e+06   
mean          1.091615e+00           1.237145e-01  2.408399e+02   
min           7.600000e-02           0.000000e+00  2.232000e+02   
25%           3.100000e-01           4.800000e-02  2.390200e+02   
50%           6.300000e-01           1.020000e-01  2.409600e+02   
75%           1.520000e+00           1.920000e-01  2.428600e+02   
max           1.112200e+01           1.390000e+00  2.541500e+02   
std           1.050655e+00           1.120142e-01  3.219643e+00   

       Global_intensity  Sub_metering_1  Sub_metering_2  Sub_metering_3  \
count      2.075259e+06    2.075259e+06    2.075259e+06    2.075259e+06   
mean       4.627759e+00    1.121923e+00    1.298520e+00    6.458447e+00   
min        2.000000e-01    0.000000e+00    0.000000e+00    0.000000e+00   
25%        1.400000e+00    0.000000e+00    0.000000e+00    0.000000e+00   
50%        2.800000e+00    0.000000e+00    0.000000e+00    1.000000e+00   
75%        6.400000e+00    0.000000e+00    1.000000e+00    1.700000e+01   
max        4.840000e+01    8.800000e+01    8.000000e+01    3.100000e+01   
std        4.416490e+00    6.114397e+00    5.785470e+00    8.384178e+00   

                            Datetime  
count                        2075259  
mean   2008-12-06 07:12:59.999994112  
min              2006-12-16 17:24:00  
25%              2007-12-12 00:18:30  
50%              2008-12-06 07:13:00  
75%              2009-12-01 14:07:30  
max              2010-11-26 21:02:00  
std                              NaN

Observation:
The dataset shows significant variability in household power consumption, with an average active power usage of 1.09 kW and peaks reaching 11.12 kW. Voltage remains fairly stable, averaging 240.8 V, while sub-metering 3 has the highest energy consumption compared to the other areas.


In [26]:
# General information of the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075259 entries, 0 to 2075258
Data columns (total 9 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   Date                   object 
 1   Time                   object 
 2   Global_active_power    object 
 3   Global_reactive_power  object 
 4   Voltage                object 
 5   Global_intensity       object 
 6   Sub_metering_1         object 
 7   Sub_metering_2         object 
 8   Sub_metering_3         float64
dtypes: float64(1), object(8)
memory usage: 142.5+ MB


Observation:
The dataset consists of 2075259 rows with 9 columns.
We can see that columns like dated", Timee", Global_active_powere", Global_reactive_power r",  Voltage t",  Global_intensityy", Sub_metering_1d", Sub_metering_2g contain string values, which are represented using the "object" datatype in this dataframe.
Only the columnSub_metering_3ar" seem to be having the "integer" datype.pe

In [28]:
# shape of the dataset
df.shape

(2075259, 9)

Observation:
The number of rows in a datset are 2075259 and Number of columns are 9

In [34]:
# Summary statistics for object (categorical) columns
df.describe(include='object')


Date      Time Global_active_power Global_reactive_power  \
count     2075259   2075259             2075259               2075259   
unique       1442      1440                6534                   896   
top     6/12/2008  17:24:00                   ?                 0.000   
freq         1440      1442               25979                472786   

        Voltage Global_intensity Sub_metering_1 Sub_metering_2  
count   2075259          2075259        2075259        2075259  
unique     5168              377            153            145  
top           ?            1.000          0.000          0.000  
freq      25979           169406        1840611        1408274

Observation:
The dataset shows 2,075,259 entries, with some missing values in the Global_active_power and Voltage columns, indicated by the "?" symbol. The most frequent values for both Sub_metering_1 and Sub_metering_2 are 0, suggesting that appliances in these categories were often not in use.

In [ ]:

df['Datetime'] = pd.to_datetime(df['Date'] + ' ' + df['Time'], format='%d/%m/%Y %H:%M:%S')
df.drop(['Date', 'Time'], axis=1, inplace=True)

In [ ]:
# Converting the columns that have object data type to float
cols_convert = ['Global_active_power', 'Global_reactive_power', 'Voltage', 'Global_intensity', 'Sub_metering_1', 'Sub_metering_2']

for col in cols_convert:
    df[col] = pd.to_numeric(df[col], errors='coerce')

In [77]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075259 entries, 0 to 2075258
Data columns (total 8 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   Global_active_power    float64       
 1   Global_reactive_power  float64       
 2   Voltage                float64       
 3   Global_intensity       float64       
 4   Sub_metering_1         float64       
 5   Sub_metering_2         float64       
 6   Sub_metering_3         float64       
 7   Datetime               datetime64[ns]
dtypes: datetime64[ns](1), float64(7)
memory usage: 126.7 MB


In [ ]:
Observation:
After converting the data type of the all the columns to float

In [38]:
# Unique values in each column
df.nunique()

Date                     1442
Time                     1440
Global_active_power      6534
Global_reactive_power     896
Voltage                  5168
Global_intensity          377
Sub_metering_1            153
Sub_metering_2            145
Sub_metering_3             32
dtype: int64

Observation:
This shows the number of unique values for each column.

In [40]:
# Count of missing values in each column
df.isnull().sum()


Date                         0
Time                         0
Global_active_power          0
Global_reactive_power        0
Voltage                      0
Global_intensity             0
Sub_metering_1               0
Sub_metering_2               0
Sub_metering_3           25979
dtype: int64

This gives the total number of missing (null) values for each column.

In [79]:
# Checking for the null values
df.isnull().any()

Global_active_power       True
Global_reactive_power     True
Voltage                   True
Global_intensity          True
Sub_metering_1            True
Sub_metering_2            True
Sub_metering_3            True
Datetime                 False
dtype: bool

Observation:
We have null values for every columns except for datetime

In [85]:
# Finding the percentage of null values
df.isnull().sum() * 100 / len(df)

Global_active_power      1.251844
Global_reactive_power    1.251844
Voltage                  1.251844
Global_intensity         1.251844
Sub_metering_1           1.251844
Sub_metering_2           1.251844
Sub_metering_3           1.251844
Datetime                 0.000000
dtype: float64

Observation:
As we can see that every columns have missing values of 1.25 % 

Dealing with missing values:
We can deal with missing values by filling the mean, median of the column or delete the entire row which have missing values


In [93]:
# filling with mean values
null_columns = ['Global_active_power', 'Global_reactive_power', 'Voltage', 'Global_intensity', 'Sub_metering_1', 'Sub_metering_2','Sub_metering_3']

for col in null_columns:
    df[col].fillna(df[col].mean())

In [95]:
# Checking for null values after filling it with mean values
df.isnull().sum()

Global_active_power      0
Global_reactive_power    0
Voltage                  0
Global_intensity         0
Sub_metering_1           0
Sub_metering_2           0
Sub_metering_3           0
Datetime                 0
dtype: int64

Observation:
After the filling it mean values null values are 0